<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Computes auto-correlation between the strands for a file of mapped sequences — gcompute_strands_autocorr • misha</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Computes auto-correlation between the strands for a file of mapped sequences — gcompute_strands_autocorr"><meta property="og:description" content="Calculates auto-correlation between plus and minus strands for the given
chromosome in a file of mapped sequences."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Computes auto-correlation between the strands for a file of mapped sequences</h1>
    <small class="dont-index">Source: <a href="https://github.com/tanaylab/misha/blob/master/R/compute.R" class="external-link"><code>R/compute.R</code></a></small>
    <div class="hidden name"><code>gcompute_strands_autocorr.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Calculates auto-correlation between plus and minus strands for the given
chromosome in a file of mapped sequences.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gcompute_strands_autocorr</span><span class="op">(</span></span>
<span>  file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrom <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  binsize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxread <span class="op">=</span> <span class="fl">400</span>,</span>
<span>  cols.order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">9</span>, <span class="fl">11</span>, <span class="fl">13</span>, <span class="fl">14</span><span class="op">)</span>,</span>
<span>  min.coord <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  max.coord <span class="op">=</span> <span class="fl">3e+08</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>the name of the file containing mapped sequences</p></dd>


<dt id="arg-chrom">chrom<a class="anchor" aria-label="anchor" href="#arg-chrom"></a></dt>
<dd><p>chromosome for which the auto-correlation is computed</p></dd>


<dt id="arg-binsize">binsize<a class="anchor" aria-label="anchor" href="#arg-binsize"></a></dt>
<dd><p>calculate the auto-correlation for bins in the range of
[-maxread, maxread]</p></dd>


<dt id="arg-maxread">maxread<a class="anchor" aria-label="anchor" href="#arg-maxread"></a></dt>
<dd><p>maximal length of the sequence used for statistics</p></dd>


<dt id="arg-cols-order">cols.order<a class="anchor" aria-label="anchor" href="#arg-cols-order"></a></dt>
<dd><p>order of sequence, chromosome, coordinate and strand
columns in file</p></dd>


<dt id="arg-min-coord">min.coord<a class="anchor" aria-label="anchor" href="#arg-min-coord"></a></dt>
<dd><p>minimal coordinate used for statistics</p></dd>


<dt id="arg-max-coord">max.coord<a class="anchor" aria-label="anchor" href="#arg-max-coord"></a></dt>
<dd><p>maximal coordinate used for statistics</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Statistics for each strand and auto-correlation by given bins.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function calculates auto-correlation between plus and minus strands for
the given chromosome in a file of mapped sequences. Each line in the file
describes one read. Each column is separated by a TAB character.</p>
<p>The following columns must be presented in the file: sequence, chromosome,
coordinate and strand. The position of these columns are controlled by
'cols.order' argument accordingly. The default value of 'cols.order' is a
vector (9,11,13,14) meaning that sequence is expected to be found at column
number 9, chromosome - at column 11, coordinate - at column 13 and strand -
at column 14. The first column should be referenced by 1 and not by 0.</p>
<p>Coordinates that are not in [min.coord, max.coord] range are ignored.</p>
<p>gcompute_strands_autocorr outputs the total statistics and the
auto-correlation given by bins. The size of the bin is indicated by
'binsize' parameter. Statistics is calculated for bins in the range of
[-maxread, maxread].</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="gdb.init.html">gdb.init_examples</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">gcompute_strands_autocorr</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">.misha</span><span class="op">$</span><span class="va">GROOT</span>, <span class="st">"reads"</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="st">"chr1"</span>, <span class="fl">50</span>,</span></span>
<span class="r-in"><span>    maxread <span class="op">=</span> <span class="fl">300</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [[1]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Forward mean Forward stdev  Reverse mean Reverse stdev </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.007209372   0.193899377   0.003404426   0.109562905 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [[2]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    bin         corr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   -6 -0.001155318</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   -5  0.003557982</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   -4 -0.001155318</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   -3  0.012984580</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   -2 -0.001155318</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   -1 -0.001155318</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7    0  0.031837776</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8    1  0.045977673</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9    2  0.196803243</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10   3  0.385335205</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11   4  0.055404271</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12   5 -0.001155318</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

