<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Initializes connection with Genomic Database — gdb.init • misha</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Initializes connection with Genomic Database — gdb.init"><meta property="og:description" content="Initializes connection with Genomic Database: loads the list of tracks,
intervals, etc."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Database-Formats.html">Database Formats and Multi-Contig Support</a>
    </li>
    <li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Initializes connection with Genomic Database</h1>
    <small class="dont-index">Source: <a href="https://github.com/tanaylab/misha/blob/master/R/db-create.R" class="external-link"><code>R/db-create.R</code></a>, <a href="https://github.com/tanaylab/misha/blob/master/R/db-root.R" class="external-link"><code>R/db-root.R</code></a></small>
    <div class="hidden name"><code>gdb.init.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Initializes connection with Genomic Database: loads the list of tracks,
intervals, etc.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gdb.init</span><span class="op">(</span>groot <span class="op">=</span> <span class="cn">NULL</span>, dir <span class="op">=</span> <span class="cn">NULL</span>, rescan <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gdb.init_examples</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gsetroot</span><span class="op">(</span>groot <span class="op">=</span> <span class="cn">NULL</span>, dir <span class="op">=</span> <span class="cn">NULL</span>, rescan <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-groot">groot<a class="anchor" aria-label="anchor" href="#arg-groot"></a></dt>
<dd><p>the root directory of the Genomic Database</p></dd>


<dt id="arg-dir">dir<a class="anchor" aria-label="anchor" href="#arg-dir"></a></dt>
<dd><p>the current working directory inside the Genomic Database</p></dd>


<dt id="arg-rescan">rescan<a class="anchor" aria-label="anchor" href="#arg-rescan"></a></dt>
<dd><p>indicates whether the file structure should be rescanned</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>None.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>'gdb.init' initializes the connection with the Genomic Database. It is
typically called first prior to any other function. When the package is
attached it internally calls to 'gdb.init.examples' which opens the
connection with the database located at 'PKGDIR/trackdb/test' directory,
where 'PKGDIR' is the directory where the package is installed.</p>
<p>The current working directory inside the Genomic Database is set to 'dir'.
If 'dir' is 'NULL', the current working directory is set to 'GROOT/tracks'.</p>
<p>If 'rescan' is 'TRUE', the list of tracks and intervals is achieved by
rescanning directory structure under the current current working directory.
Otherwise 'gdb.init' attempts to use the cached list that resides in
'groot/.db.cache' file.</p>
<p>Upon completion the connection is established with the database. If
auto-completion mode is switched on (see 'gset_input_method') the list of
tracks and intervals sets is loaded and added as variables to the global
environment allowing auto-completion of object names with &lt;TAB&gt; key. Also a
few variables are defined at an environment called <code>.misha</code>, and can be
accessed using <code>.misha$variable</code>, e.g. <code>.misha$ALLGENOME</code>.
These variables should not be modified by user.</p>
<table class="table table"><tr><td>GROOT</td><td>Root directory of Genomic Database</td></tr><tr><td>GWD</td><td>Current working directory inside Genomic Database</td></tr><tr><td>GTRACKS</td><td>List of
all available tracks</td></tr><tr><td>GINTERVS</td><td>List of all available intervals</td></tr><tr><td>GVTRACKS</td><td>List of all available virtual tracks</td></tr><tr><td>ALLGENOME</td><td>List of
all chromosomes and their sizes</td></tr><tr><td>GITERATOR.INTERVALS</td><td>A set of
iterator intervals for which the track expression is evaluated</td></tr></table><p>When option 'gmulticontig.indexed_format' is set to TRUE, the function
loads a database with "indexed" track format.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="gdb.reload.html">gdb.reload</a></code>, <code><a href="gdb.create.html">gdb.create</a></code>,
<code><a href="gdir.cd.html">gdir.cd</a></code>, <code><a href="gtrack.ls.html">gtrack.ls</a></code>, <code><a href="gintervals.ls.html">gintervals.ls</a></code>,
<code><a href="gvtrack.ls.html">gvtrack.ls</a></code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

