<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Score DNA sequences with a k-mer over a region of interest — gseq.kmer • misha</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Score DNA sequences with a k-mer over a region of interest — gseq.kmer"><meta property="og:description" content="Counts exact matches of a k-mer in DNA sequences over a specified region of interest
(ROI). The ROI is defined by start_pos and end_pos (1-based, inclusive),
with optional extension controlled by extend."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Database-Formats.html">Database Formats and Multi-Contig Support</a>
    </li>
    <li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Score DNA sequences with a k-mer over a region of interest</h1>
    <small class="dont-index">Source: <a href="https://github.com/tanaylab/misha/blob/master/R/sequence.R" class="external-link"><code>R/sequence.R</code></a></small>
    <div class="hidden name"><code>gseq.kmer.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Counts exact matches of a k-mer in DNA sequences over a specified region of interest
(ROI). The ROI is defined by <code>start_pos</code> and <code>end_pos</code> (1-based, inclusive),
with optional extension controlled by <code>extend</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gseq.kmer</span><span class="op">(</span></span>
<span>  <span class="va">seqs</span>,</span>
<span>  <span class="va">kmer</span>,</span>
<span>  mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"count"</span>, <span class="st">"frac"</span><span class="op">)</span>,</span>
<span>  strand <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  start_pos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  end_pos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  skip_gaps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gap_chars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">"."</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-seqs">seqs<a class="anchor" aria-label="anchor" href="#arg-seqs"></a></dt>
<dd><p>character vector of DNA sequences (A/C/G/T/N; case-insensitive)</p></dd>


<dt id="arg-kmer">kmer<a class="anchor" aria-label="anchor" href="#arg-kmer"></a></dt>
<dd><p>single character string containing the k-mer to search for (A/C/G/T only)</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>character; one of "count" or "frac"</p></dd>


<dt id="arg-strand">strand<a class="anchor" aria-label="anchor" href="#arg-strand"></a></dt>
<dd><p>integer; 1=forward, -1=reverse, 0=both strands (default: 0)</p></dd>


<dt id="arg-start-pos">start_pos<a class="anchor" aria-label="anchor" href="#arg-start-pos"></a></dt>
<dd><p>integer or NULL; 1-based inclusive start of ROI (default: 1)</p></dd>


<dt id="arg-end-pos">end_pos<a class="anchor" aria-label="anchor" href="#arg-end-pos"></a></dt>
<dd><p>integer or NULL; 1-based inclusive end of ROI (default: sequence length)</p></dd>


<dt id="arg-extend">extend<a class="anchor" aria-label="anchor" href="#arg-extend"></a></dt>
<dd><p>logical or integer; extension of allowed window starts (default: FALSE)</p></dd>


<dt id="arg-skip-gaps">skip_gaps<a class="anchor" aria-label="anchor" href="#arg-skip-gaps"></a></dt>
<dd><p>logical; if TRUE, treat gap characters as holes and skip them while
scanning. Windows are k consecutive non-gap bases (default: TRUE)</p></dd>


<dt id="arg-gap-chars">gap_chars<a class="anchor" aria-label="anchor" href="#arg-gap-chars"></a></dt>
<dd><p>character vector; which characters count as gaps (default: c("-", "."))</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Numeric vector with counts (for "count" mode) or fractions (for "frac" mode).
  Returns 0 when sequence is too short or ROI is invalid.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function counts k-mer occurrences in DNA sequences directly without requiring
a genomics database. For detailed documentation on k-mer counting parameters, see
<code><a href="gvtrack.create.html">gvtrack.create</a></code> (functions "kmer.count" and "kmer.frac").</p>
<p>The ROI (region of interest) is defined by <code>start_pos</code> and <code>end_pos</code>.
The <code>extend</code> parameter controls whether k-mer matches can extend beyond the ROI boundaries.
For palindromic k-mers, use <code>strand=1</code> or <code>-1</code> to avoid double counting.</p>
<p>When <code>skip_gaps=TRUE</code>, characters specified in <code>gap_chars</code> are treated as gaps.
Windows are defined as k consecutive non-gap bases. The <code>frac</code> denominator counts the
number of possible logical starts (non-gap windows) in the region. <code>start_pos</code> and
<code>end_pos</code> are interpreted as physical coordinates on the full sequence.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="gvtrack.create.html">gvtrack.create</a></code> for detailed k-mer parameter documentation</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Example sequences</span></span></span>
<span class="r-in"><span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CGCGCGCGCG"</span>, <span class="st">"ATATATATAT"</span>, <span class="st">"ACGTACGTACGT"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Count CG dinucleotides on both strands</span></span></span>
<span class="r-in"><span><span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"CG"</span>, mode <span class="op">=</span> <span class="st">"count"</span>, strand <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Count on forward strand only</span></span></span>
<span class="r-in"><span><span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"CG"</span>, mode <span class="op">=</span> <span class="st">"count"</span>, strand <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get CG fraction</span></span></span>
<span class="r-in"><span><span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"CG"</span>, mode <span class="op">=</span> <span class="st">"frac"</span>, strand <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Count in a specific region</span></span></span>
<span class="r-in"><span><span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"CG"</span>, mode <span class="op">=</span> <span class="st">"count"</span>, start_pos <span class="op">=</span> <span class="fl">2</span>, end_pos <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Allow k-mer to extend beyond ROI boundaries</span></span></span>
<span class="r-in"><span><span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"CG"</span>, mode <span class="op">=</span> <span class="st">"count"</span>, start_pos <span class="op">=</span> <span class="fl">2</span>, end_pos <span class="op">=</span> <span class="fl">8</span>, extend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Calculate GC content by summing G and C fractions</span></span></span>
<span class="r-in"><span><span class="va">g_frac</span> <span class="op">&lt;-</span> <span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"G"</span>, mode <span class="op">=</span> <span class="st">"frac"</span>, strand <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">c_frac</span> <span class="op">&lt;-</span> <span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"C"</span>, mode <span class="op">=</span> <span class="st">"frac"</span>, strand <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gc_content</span> <span class="op">&lt;-</span> <span class="va">g_frac</span> <span class="op">+</span> <span class="va">c_frac</span></span></span>
<span class="r-in"><span><span class="va">gc_content</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compare AT counts on different strands</span></span></span>
<span class="r-in"><span><span class="va">at_forward</span> <span class="op">&lt;-</span> <span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"AT"</span>, mode <span class="op">=</span> <span class="st">"count"</span>, strand <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">at_reverse</span> <span class="op">&lt;-</span> <span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"AT"</span>, mode <span class="op">=</span> <span class="st">"count"</span>, strand <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">at_both</span> <span class="op">&lt;-</span> <span class="fu">gseq.kmer</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"AT"</span>, mode <span class="op">=</span> <span class="st">"count"</span>, strand <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>forward <span class="op">=</span> <span class="va">at_forward</span>, reverse <span class="op">=</span> <span class="va">at_reverse</span>, both <span class="op">=</span> <span class="va">at_both</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

