<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate random genome sequences — gsynth.random • misha</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate random genome sequences — gsynth.random"><meta property="og:description" content="Generates random DNA sequences based on nucleotide probabilities without
using a trained Markov model. Each nucleotide is sampled independently
according to the specified probabilities."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.4.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Database-Formats.html">Database Formats and Multi-Contig Support</a>
    </li>
    <li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
    <li>
      <a href="../articles/Misha-Basics.html">Misha Basics (Short Guide)</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate random genome sequences</h1>
    <small class="dont-index">Source: <a href="https://github.com/tanaylab/misha/blob/master/R/synth.R" class="external-link"><code>R/synth.R</code></a></small>
    <div class="hidden name"><code>gsynth.random.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generates random DNA sequences based on nucleotide probabilities without
using a trained Markov model. Each nucleotide is sampled independently
according to the specified probabilities.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gsynth.random</span><span class="op">(</span></span>
<span>  intervals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output_format <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"misha"</span>, <span class="st">"fasta"</span>, <span class="st">"vector"</span><span class="op">)</span>,</span>
<span>  nuc_probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fl">0.25</span>, C <span class="op">=</span> <span class="fl">0.25</span>, G <span class="op">=</span> <span class="fl">0.25</span>, T <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span>,</span>
<span>  mask_copy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_samples <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  iterator <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-intervals">intervals<a class="anchor" aria-label="anchor" href="#arg-intervals"></a></dt>
<dd><p>Genomic intervals to sample. If NULL, uses all chromosomes.</p></dd>


<dt id="arg-output-path">output_path<a class="anchor" aria-label="anchor" href="#arg-output-path"></a></dt>
<dd><p>Path to the output file (ignored when output_format = "vector")</p></dd>


<dt id="arg-output-format">output_format<a class="anchor" aria-label="anchor" href="#arg-output-format"></a></dt>
<dd><p>Output format:</p><ul><li><p>"misha": .seq binary format (default)</p></li>
<li><p>"fasta": FASTA text format</p></li>
<li><p>"vector": Return sequences as a character vector (does not write to file)</p></li>
</ul></dd>


<dt id="arg-nuc-probs">nuc_probs<a class="anchor" aria-label="anchor" href="#arg-nuc-probs"></a></dt>
<dd><p>Nucleotide probabilities. Can be specified as:</p><ul><li><p>A named vector: <code>c(A = 0.3, C = 0.2, G = 0.2, T = 0.3)</code></p></li>
<li><p>An unnamed vector in A, C, G, T order: <code>c(0.3, 0.2, 0.2, 0.3)</code></p></li>
</ul><p>Probabilities are automatically normalized to sum to 1. Default is uniform
(0.25 each).</p></dd>


<dt id="arg-mask-copy">mask_copy<a class="anchor" aria-label="anchor" href="#arg-mask-copy"></a></dt>
<dd><p>Optional intervals to copy from the original genome instead of
random sampling. Use this to preserve specific regions exactly as they
appear in the reference.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed for reproducibility. If NULL, uses current random state.</p></dd>


<dt id="arg-n-samples">n_samples<a class="anchor" aria-label="anchor" href="#arg-n-samples"></a></dt>
<dd><p>Number of samples to generate per interval. Default is 1.</p></dd>


<dt id="arg-iterator">iterator<a class="anchor" aria-label="anchor" href="#arg-iterator"></a></dt>
<dd><p>Iterator for position resolution. Default is 1 (base-pair resolution).
Larger values may speed up processing but are typically not needed for
random sampling.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>When output_format is "misha" or "fasta", returns invisible NULL and
        writes the random sequences to output_path. When output_format is "vector",
        returns a character vector of sequences (length = n_intervals * n_samples).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Unlike <code><a href="gsynth.sample.html">gsynth.sample</a></code> which uses a trained Markov model to generate
sequences that preserve k-mer statistics, <code>gsynth.random</code> generates purely
random sequences where each nucleotide is sampled independently. This is useful
for generating baseline random sequences or sequences with specific GC content.</p>
<p><strong>Nucleotide ordering:</strong> When using an unnamed vector for <code>nuc_probs</code>,
the order is A, C, G, T. Named vectors can be in any order.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="gsynth.sample.html">gsynth.sample</a></code>, <code><a href="gsynth.train.html">gsynth.train</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="gdb.init.html">gdb.init_examples</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate random sequences with uniform nucleotide probabilities</span></span></span>
<span class="r-in"><span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">gsynth.random</span><span class="op">(</span></span></span>
<span class="r-in"><span>    intervals <span class="op">=</span> <span class="fu"><a href="gintervals.html">gintervals</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    output_format <span class="op">=</span> <span class="st">"vector"</span>,</span></span>
<span class="r-in"><span>    seed <span class="op">=</span> <span class="fl">42</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Setting up random sampling positions...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating random sequences (1 samples per interval)...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generated 1 random sequence(s)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate GC-rich sequences (60% GC)</span></span></span>
<span class="r-in"><span><span class="va">gc_rich</span> <span class="op">&lt;-</span> <span class="fu">gsynth.random</span><span class="op">(</span></span></span>
<span class="r-in"><span>    intervals <span class="op">=</span> <span class="fu"><a href="gintervals.html">gintervals</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    output_format <span class="op">=</span> <span class="st">"vector"</span>,</span></span>
<span class="r-in"><span>    nuc_probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fl">0.2</span>, C <span class="op">=</span> <span class="fl">0.3</span>, G <span class="op">=</span> <span class="fl">0.3</span>, T <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    seed <span class="op">=</span> <span class="fl">42</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Setting up random sampling positions...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating random sequences (1 samples per interval)...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generated 1 random sequence(s)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate AT-rich sequences</span></span></span>
<span class="r-in"><span><span class="va">at_rich</span> <span class="op">&lt;-</span> <span class="fu">gsynth.random</span><span class="op">(</span></span></span>
<span class="r-in"><span>    intervals <span class="op">=</span> <span class="fu"><a href="gintervals.html">gintervals</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    output_format <span class="op">=</span> <span class="st">"vector"</span>,</span></span>
<span class="r-in"><span>    nuc_probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fl">0.35</span>, C <span class="op">=</span> <span class="fl">0.15</span>, G <span class="op">=</span> <span class="fl">0.15</span>, T <span class="op">=</span> <span class="fl">0.35</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    seed <span class="op">=</span> <span class="fl">42</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Setting up random sampling positions...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating random sequences (1 samples per interval)...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generated 1 random sequence(s)</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

