<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Iteratively replace a k-mer in the genome — gsynth.replace_kmer • misha</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Iteratively replace a k-mer in the genome — gsynth.replace_kmer"><meta property="og:description" content="Performs an iterative replacement of a target k-mer with a
replacement sequence. This is useful for creating synthetic genomes
with specific motifs removed (e.g., creating a CpG-null genome by iteratively
swapping CG to GC)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Database-Formats.html">Database Formats and Multi-Contig Support</a>
    </li>
    <li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Iteratively replace a k-mer in the genome</h1>
    <small class="dont-index">Source: <a href="https://github.com/tanaylab/misha/blob/master/R/synth.R" class="external-link"><code>R/synth.R</code></a></small>
    <div class="hidden name"><code>gsynth.replace_kmer.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs an iterative replacement of a <code>target</code> k-mer with a
<code>replacement</code> sequence. This is useful for creating synthetic genomes
with specific motifs removed (e.g., creating a CpG-null genome by iteratively
swapping CG to GC).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gsynth.replace_kmer</span><span class="op">(</span></span>
<span>  <span class="va">target</span>,</span>
<span>  <span class="va">replacement</span>,</span>
<span>  output_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output_format <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"misha"</span>, <span class="st">"fasta"</span>, <span class="st">"vector"</span><span class="op">)</span>,</span>
<span>  intervals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  check_composition <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-target">target<a class="anchor" aria-label="anchor" href="#arg-target"></a></dt>
<dd><p>The k-mer sequence to remove (e.g., "CG").</p></dd>


<dt id="arg-replacement">replacement<a class="anchor" aria-label="anchor" href="#arg-replacement"></a></dt>
<dd><p>The replacement sequence (e.g., "GC").</p></dd>


<dt id="arg-output-path">output_path<a class="anchor" aria-label="anchor" href="#arg-output-path"></a></dt>
<dd><p>Path to the output file (ignored when output_format = "vector").</p></dd>


<dt id="arg-output-format">output_format<a class="anchor" aria-label="anchor" href="#arg-output-format"></a></dt>
<dd><p>Output format:</p><ul><li><p>"misha": .seq binary format (default)</p></li>
<li><p>"fasta": FASTA text format</p></li>
<li><p>"vector": Return sequences as a character vector (does not write to file)</p></li>
</ul></dd>


<dt id="arg-intervals">intervals<a class="anchor" aria-label="anchor" href="#arg-intervals"></a></dt>
<dd><p>Genomic intervals to process. If NULL, uses all chromosomes.</p></dd>


<dt id="arg-check-composition">check_composition<a class="anchor" aria-label="anchor" href="#arg-check-composition"></a></dt>
<dd><p>Logical. If TRUE (default), ensures target and
replacement have the same nucleotide composition (preserving exact
base counts).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>When output_format is "misha" or "fasta", returns invisible NULL and
        writes to output_path. When output_format is "vector", returns a
        character vector of modified sequences.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Bubble Sort / Iterative Logic:</strong> The function scans the sequence and
replaces occurrences of <code>target</code> with <code>replacement</code>. If a replacement
creates a new instance of <code>target</code> (e.g., removing "CG" with "GC" in
the sequence "CCG" -&gt; "CGC"), the new instance is also replaced. This continues
until the sequence is free of the <code>target</code> k-mer.</p>
<p>When <code>target</code> and <code>replacement</code> are permutations of each other
(e.g., "CG" and "GC"), this acts as a "bubble sort" of nucleotides, moving
bases locally without altering the total GC content or base counts of the genome.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Robust removal of all CpG dinucleotides (preserving GC%)</span></span></span>
<span class="r-in"><span><span class="fu">gsynth.replace_kmer</span><span class="op">(</span></span></span>
<span class="r-in"><span>    target <span class="op">=</span> <span class="st">"CG"</span>,</span></span>
<span class="r-in"><span>    replacement <span class="op">=</span> <span class="st">"GC"</span>,</span></span>
<span class="r-in"><span>    output_path <span class="op">=</span> <span class="st">"genome_no_cpg.seq"</span>,</span></span>
<span class="r-in"><span>    output_format <span class="op">=</span> <span class="st">"misha"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

