<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Train a stratified Markov-5 model from genome sequences — gsynth.train • misha</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Train a stratified Markov-5 model from genome sequences — gsynth.train"><meta property="og:description" content="Computes a 5th-order Markov model optionally stratified by bins of one or more
track expressions (e.g., GC content and CG dinucleotide frequency). This model
can be used to generate synthetic genomes that preserve the k-mer statistics of
the original genome within each stratification bin. When called with no dimension
specifications, trains a single unstratified model."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.4.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Database-Formats.html">Database Formats and Multi-Contig Support</a>
    </li>
    <li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
    <li>
      <a href="../articles/Misha-Basics.html">Misha Basics (Short Guide)</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Train a stratified Markov-5 model from genome sequences</h1>
    <small class="dont-index">Source: <a href="https://github.com/tanaylab/misha/blob/v5.4.5/R/synth.R" class="external-link"><code>R/synth.R</code></a></small>
    <div class="hidden name"><code>gsynth.train.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Computes a 5th-order Markov model optionally stratified by bins of one or more
track expressions (e.g., GC content and CG dinucleotide frequency). This model
can be used to generate synthetic genomes that preserve the k-mer statistics of
the original genome within each stratification bin. When called with no dimension
specifications, trains a single unstratified model.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gsynth.train</span><span class="op">(</span></span>
<span>  <span class="va">...</span>,</span>
<span>  mask <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intervals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  iterator <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pseudocount <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  min_obs <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Zero or more dimension specifications. Each specification is a list
     containing:</p><dl><dt>expr</dt>
<dd><p>Track expression for this dimension (required)</p></dd>

  <dt>breaks</dt>
<dd><p>Numeric vector of bin boundaries for this dimension (required)</p></dd>

  <dt>bin_merge</dt>
<dd><p>Optional list of merge specifications for merging sparse bins.
        Each specification is a named list with 'from' and 'to' elements.</p></dd>


</dl><p>If no dimensions are provided, trains an unstratified model with a single bin.</p></dd>


<dt id="arg-mask">mask<a class="anchor" aria-label="anchor" href="#arg-mask"></a></dt>
<dd><p>Optional intervals to exclude from training. Regions in the mask
will not contribute to k-mer counts. Can be computed using <code><a href="gscreen.html">gscreen()</a></code>.</p></dd>


<dt id="arg-intervals">intervals<a class="anchor" aria-label="anchor" href="#arg-intervals"></a></dt>
<dd><p>Genomic intervals to process. If NULL, uses all chromosomes.</p></dd>


<dt id="arg-iterator">iterator<a class="anchor" aria-label="anchor" href="#arg-iterator"></a></dt>
<dd><p>Iterator for track evaluation, determines the resolution at which
track values are computed.</p></dd>


<dt id="arg-pseudocount">pseudocount<a class="anchor" aria-label="anchor" href="#arg-pseudocount"></a></dt>
<dd><p>Pseudocount added to all k-mer counts to avoid zero probabilities.
Default is 1.</p></dd>


<dt id="arg-min-obs">min_obs<a class="anchor" aria-label="anchor" href="#arg-min-obs"></a></dt>
<dd><p>Minimum number of observations (6-mers) required per bin. Bins with
fewer observations will be marked as NA (not learned) and a warning will be
issued. Default is 0 (no minimum). During sampling, NA bins will fall back
to uniform sampling unless merged via <code>bin_merge</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>gsynth.model</code> object containing:</p><dl><dt>n_dims</dt>
<dd><p>Number of stratification dimensions</p></dd>

    <dt>dim_specs</dt>
<dd><p>List of dimension specifications (expr, breaks, num_bins, bin_map)</p></dd>

    <dt>dim_sizes</dt>
<dd><p>Vector of bin counts per dimension</p></dd>

    <dt>total_bins</dt>
<dd><p>Total number of bins (product of dim_sizes)</p></dd>

    <dt>total_kmers</dt>
<dd><p>Total number of valid 6-mers counted</p></dd>

    <dt>per_bin_kmers</dt>
<dd><p>Number of 6-mers counted per bin</p></dd>

    <dt>total_masked</dt>
<dd><p>Number of positions skipped due to mask</p></dd>

    <dt>total_n</dt>
<dd><p>Number of positions skipped due to N bases</p></dd>

    <dt>model_data</dt>
<dd><p>Internal model data (counts and CDFs)</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Strand symmetry:</strong> The training process counts both the forward strand
6-mer and its reverse complement for each position, ensuring strand-symmetric
transition probabilities. This means the reported total_kmers is approximately
double the number of genomic positions processed.</p>
<p><strong>N bases:</strong> Positions where the 6-mer contains any N (unknown) bases are
skipped during training and counted in <code>total_n</code>. The model only learns
from valid A/C/G/T sequences.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="gsynth.sample.html">gsynth.sample</a></code>, <code><a href="gsynth.save.html">gsynth.save</a></code>,
         <code><a href="gsynth.load.html">gsynth.load</a></code>, <code><a href="gsynth.bin_map.html">gsynth.bin_map</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="gdb.init.html">gdb.init_examples</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create virtual tracks for stratification</span></span></span>
<span class="r-in"><span><span class="fu"><a href="gvtrack.create.html">gvtrack.create</a></span><span class="op">(</span><span class="st">"g_frac"</span>, <span class="cn">NULL</span>, <span class="st">"kmer.frac"</span>, kmer <span class="op">=</span> <span class="st">"G"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="gvtrack.create.html">gvtrack.create</a></span><span class="op">(</span><span class="st">"c_frac"</span>, <span class="cn">NULL</span>, <span class="st">"kmer.frac"</span>, kmer <span class="op">=</span> <span class="st">"C"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="gvtrack.create.html">gvtrack.create</a></span><span class="op">(</span><span class="st">"cg_frac"</span>, <span class="cn">NULL</span>, <span class="st">"kmer.frac"</span>, kmer <span class="op">=</span> <span class="st">"CG"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>kmer sequence 'CG' is palindromic, please set strand to 1 or -1 to avoid double counting</span>
<span class="r-in"><span><span class="fu"><a href="gvtrack.create.html">gvtrack.create</a></span><span class="op">(</span><span class="st">"masked_frac"</span>, <span class="cn">NULL</span>, <span class="st">"masked.frac"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Define repeat mask</span></span></span>
<span class="r-in"><span><span class="va">repeats</span> <span class="op">&lt;-</span> <span class="fu"><a href="gscreen.html">gscreen</a></span><span class="op">(</span><span class="st">"masked_frac &gt; 0.5"</span>,</span></span>
<span class="r-in"><span>    intervals <span class="op">=</span> <span class="fu"><a href="gintervals.all.html">gintervals.all</a></span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    iterator <span class="op">=</span> <span class="fl">100</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Train unstratified model (no stratification)</span></span></span>
<span class="r-in"><span><span class="va">model_0d</span> <span class="op">&lt;-</span> <span class="fu">gsynth.train</span><span class="op">(</span></span></span>
<span class="r-in"><span>    mask <span class="op">=</span> <span class="va">repeats</span>,</span></span>
<span class="r-in"><span>    intervals <span class="op">=</span> <span class="fu"><a href="gintervals.all.html">gintervals.all</a></span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    iterator <span class="op">=</span> <span class="fl">200</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Setting up iterator positions...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Training Markov model...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Trained unstratified model: 835,310 6-mers (no stratification)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Train model with 2D stratification (GC content and CG dinucleotide)</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu">gsynth.train</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>        expr <span class="op">=</span> <span class="st">"g_frac + c_frac"</span>,</span></span>
<span class="r-in"><span>        breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.025</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        bin_merge <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.7</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.675</span>, <span class="fl">0.7</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>        expr <span class="op">=</span> <span class="st">"cg_frac"</span>,</span></span>
<span class="r-in"><span>        breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.01</span>, <span class="fl">0.02</span>, <span class="fl">0.03</span>, <span class="fl">0.04</span>, <span class="fl">0.2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        bin_merge <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.04</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.03</span>, <span class="fl">0.04</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>,</span></span>
<span class="r-in"><span>    mask <span class="op">=</span> <span class="va">repeats</span>,</span></span>
<span class="r-in"><span>    intervals <span class="op">=</span> <span class="fu"><a href="gintervals.all.html">gintervals.all</a></span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    iterator <span class="op">=</span> <span class="fl">200</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting track values...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Training Markov model...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Trained model: 835,310 6-mers across 200 bins (2 dimensions)</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

