<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ misha</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Database-Formats.html">Database Formats and Multi-Contig Support</a>
    </li>
    <li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="track-functions">Track functions <a href="#track-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to manipulate and create tracks</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gtrack.2d.create.html">gtrack.2d.create()</a></code> </p>
        </td>
        <td><p>Creates a 'Rectangles' track from intervals and values</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.2d.import.html">gtrack.2d.import()</a></code> </p>
        </td>
        <td><p>Creates a 2D track from tab-delimited file</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.2d.import_contacts.html">gtrack.2d.import_contacts()</a></code> </p>
        </td>
        <td><p>Creates a track from a file of inter-genomic contacts</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.array.extract.html">gtrack.array.extract()</a></code> </p>
        </td>
        <td><p>Returns values from 'Array' track</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.array.get_colnames.html">gtrack.array.get_colnames()</a></code> </p>
        </td>
        <td><p>Returns column names of array track</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.array.import.html">gtrack.array.import()</a></code> </p>
        </td>
        <td><p>Creates an array track from array tracks or files</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.array.set_colnames.html">gtrack.array.set_colnames()</a></code> </p>
        </td>
        <td><p>Sets column names of array track</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.attr.export.html">gtrack.attr.export()</a></code> </p>
        </td>
        <td><p>Returns track attributes values</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.attr.get.html">gtrack.attr.get()</a></code> </p>
        </td>
        <td><p>Returns value of a track attribute</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.attr.import.html">gtrack.attr.import()</a></code> </p>
        </td>
        <td><p>Imports track attributes values</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.attr.set.html">gtrack.attr.set()</a></code> </p>
        </td>
        <td><p>Assigns value to a track attribute</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.convert.html">gtrack.convert()</a></code> </p>
        </td>
        <td><p>Converts a track to the most current format</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.create.html">gtrack.create()</a></code> </p>
        </td>
        <td><p>Creates a track from a track expression</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.create_pwm_energy.html">gtrack.create_pwm_energy()</a></code> </p>
        </td>
        <td><p>Creates a new track from PSSM energy function</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.create_sparse.html">gtrack.create_sparse()</a></code> </p>
        </td>
        <td><p>Creates a 'Sparse' track from intervals and values</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.create_dense.html">gtrack.create_dense()</a></code> </p>
        </td>
        <td><p>Creates a 'Dense' track from intervals and values</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.create_dirs.html">gtrack.create_dirs()</a></code> </p>
        </td>
        <td><p>Create directories needed for track creation</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.exists.html">gtrack.exists()</a></code> </p>
        </td>
        <td><p>Tests for a track existence</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.import.html">gtrack.import()</a></code> </p>
        </td>
        <td><p>Creates a track from WIG / BigWig / BedGraph / BED / tab-delimited file</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.import_mappedseq.html">gtrack.import_mappedseq()</a></code> </p>
        </td>
        <td><p>Creates a track from a file of mapped sequences</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.import_set.html">gtrack.import_set()</a></code> </p>
        </td>
        <td><p>Creates one or more tracks from multiple WIG / BigWig / BedGraph / tab-delimited files on disk or FTP</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.info.html">gtrack.info()</a></code> </p>
        </td>
        <td><p>Returns information about a track</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.path.html">gtrack.path()</a></code> </p>
        </td>
        <td><p>Returns the path on disk of a track</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.liftover.html">gtrack.liftover()</a></code> </p>
        </td>
        <td><p>Imports a track from another assembly</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.lookup.html">gtrack.lookup()</a></code> </p>
        </td>
        <td><p>Creates a new track from a lookup table based on track expression</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.ls.html">gtrack.ls()</a></code> </p>
        </td>
        <td><p>Returns a list of track names</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.modify.html">gtrack.modify()</a></code> </p>
        </td>
        <td><p>Modifies track contents</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.rm.html">gtrack.rm()</a></code> </p>
        </td>
        <td><p>Deletes a track</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.smooth.html">gtrack.smooth()</a></code> </p>
        </td>
        <td><p>Creates a new track from smoothed values of track expression</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.var.get.html">gtrack.var.get()</a></code> </p>
        </td>
        <td><p>Returns value of a track variable</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.var.ls.html">gtrack.var.ls()</a></code> </p>
        </td>
        <td><p>Returns a list of track variables for a track</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.var.rm.html">gtrack.var.rm()</a></code> </p>
        </td>
        <td><p>Deletes a track variable</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.var.set.html">gtrack.var.set()</a></code> </p>
        </td>
        <td><p>Assigns value to a track variable</p></td>
      </tr><tr><td>
          <p><code><a href="gtrack.convert_to_indexed.html">gtrack.convert_to_indexed()</a></code> </p>
        </td>
        <td><p>Convert a track to indexed format</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="virtual-track-functions">Virtual track functions <a href="#virtual-track-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to manipulate and create virtual tracks</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gvtrack.array.slice.html">gvtrack.array.slice()</a></code> </p>
        </td>
        <td><p>Defines rules for a single value calculation of a virtual 'Array' track</p></td>
      </tr><tr><td>
          <p><code><a href="gvtrack.create.html">gvtrack.create()</a></code> </p>
        </td>
        <td><p>Creates a new virtual track</p></td>
      </tr><tr><td>
          <p><code><a href="gvtrack.info.html">gvtrack.info()</a></code> </p>
        </td>
        <td><p>Returns the definition of a virtual track</p></td>
      </tr><tr><td>
          <p><code><a href="gvtrack.iterator.html">gvtrack.iterator()</a></code> </p>
        </td>
        <td><p>Defines modification rules for a one-dimensional iterator in a virtual track</p></td>
      </tr><tr><td>
          <p><code><a href="gvtrack.iterator.2d.html">gvtrack.iterator.2d()</a></code> </p>
        </td>
        <td><p>Defines modification rules for a two-dimensional iterator in a virtual track</p></td>
      </tr><tr><td>
          <p><code><a href="gvtrack.ls.html">gvtrack.ls()</a></code> </p>
        </td>
        <td><p>Returns a list of virtual track names</p></td>
      </tr><tr><td>
          <p><code><a href="gvtrack.rm.html">gvtrack.rm()</a></code> </p>
        </td>
        <td><p>Deletes a virtual track</p></td>
      </tr><tr><td>
          <p><code><a href="gvtrack.filter.html">gvtrack.filter()</a></code> </p>
        </td>
        <td><p>Attach or clear a genomic mask filter on a virtual track</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="intervals-functions">Intervals functions <a href="#intervals-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to manipulate and create intervals</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gintervals.html">gintervals()</a></code> </p>
        </td>
        <td><p>Creates a set of 1D intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.2d.html">gintervals.2d()</a></code> </p>
        </td>
        <td><p>Creates a set of 2D intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.2d.all.html">gintervals.2d.all()</a></code> </p>
        </td>
        <td><p>Returns 2D intervals that cover the whole genome</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.2d.band_intersect.html">gintervals.2d.band_intersect()</a></code> </p>
        </td>
        <td><p>Intersects two-dimensional intervals with a band</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.all.html">gintervals.all()</a></code> </p>
        </td>
        <td><p>Returns 1D intervals that cover the whole genome</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.canonic.html">gintervals.canonic()</a></code> </p>
        </td>
        <td><p>Converts intervals to canonic form</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.mark_overlaps.html">gintervals.mark_overlaps()</a></code> </p>
        </td>
        <td><p>Mark overlapping intervals with a group ID</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.chrom_sizes.html">gintervals.chrom_sizes()</a></code> </p>
        </td>
        <td><p>Returns number of intervals per chromosome</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.covered_bp.html">gintervals.covered_bp()</a></code> </p>
        </td>
        <td><p>Calculate total base pairs covered by intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.coverage_fraction.html">gintervals.coverage_fraction()</a></code> </p>
        </td>
        <td><p>Calculate fraction of genomic space covered by intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.diff.html">gintervals.diff()</a></code> </p>
        </td>
        <td><p>Calculates difference of two intervals sets</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.exists.html">gintervals.exists()</a></code> </p>
        </td>
        <td><p>Tests for a named intervals set existence</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.force_range.html">gintervals.force_range()</a></code> </p>
        </td>
        <td><p>Limits intervals to chromosomal range</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.normalize.html">gintervals.normalize()</a></code> </p>
        </td>
        <td><p>Normalize intervals to a fixed size</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.annotate.html">gintervals.annotate()</a></code> </p>
        </td>
        <td><p>Annotates 1D intervals using nearest neighbors</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.import_genes.html">gintervals.import_genes()</a></code> </p>
        </td>
        <td><p>Imports genes and annotations from files</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.intersect.html">gintervals.intersect()</a></code> </p>
        </td>
        <td><p>Calculates an intersection of two sets of intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.is.bigset.html">gintervals.is.bigset()</a></code> </p>
        </td>
        <td><p>Tests for big intervals set</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.liftover.html">gintervals.liftover()</a></code> </p>
        </td>
        <td><p>Converts intervals from another assembly</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.load.html">gintervals.load()</a></code> </p>
        </td>
        <td><p>Loads a named intervals set</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.load_chain.html">gintervals.load_chain()</a></code> </p>
        </td>
        <td><p>Loads assembly conversion table from a chain file</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.as_chain.html">gintervals.as_chain()</a></code> </p>
        </td>
        <td><p>Transforms existing intervals to a chain format</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.path.html">gintervals.path()</a></code> </p>
        </td>
        <td><p>Returns the path on disk of an interval set</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.ls.html">gintervals.ls()</a></code> </p>
        </td>
        <td><p>Returns a list of named intervals sets</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.mapply.html">gintervals.mapply()</a></code> </p>
        </td>
        <td><p>Applies a function to values of track expressions</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.neighbors.html">gintervals.neighbors()</a></code> </p>
        </td>
        <td><p>Finds neighbors between two sets of intervals</p></td>
      </tr><tr><td>
          <p><code><a href="directional-neighbors.html">gintervals.neighbors.upstream()</a></code> <code><a href="directional-neighbors.html">gintervals.neighbors.downstream()</a></code> <code><a href="directional-neighbors.html">gintervals.neighbors.directional()</a></code> </p>
        </td>
        <td><p>Directional neighbor finding functions</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.quantiles.html">gintervals.quantiles()</a></code> </p>
        </td>
        <td><p>Calculates quantiles of a track expression for intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.random.html">gintervals.random()</a></code> </p>
        </td>
        <td><p>Generate random genome intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.rbind.html">gintervals.rbind()</a></code> </p>
        </td>
        <td><p>Combines several sets of intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.rm.html">gintervals.rm()</a></code> </p>
        </td>
        <td><p>Deletes a named intervals set</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.save.html">gintervals.save()</a></code> </p>
        </td>
        <td><p>Creates a named intervals set</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.summary.html">gintervals.summary()</a></code> </p>
        </td>
        <td><p>Calculates summary statistics of track expression for intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.union.html">gintervals.union()</a></code> </p>
        </td>
        <td><p>Calculates a union of two sets of intervals</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.update.html">gintervals.update()</a></code> </p>
        </td>
        <td><p>Updates a named intervals set</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.2d.convert_to_indexed.html">gintervals.2d.convert_to_indexed()</a></code> </p>
        </td>
        <td><p>Convert 2D interval set to indexed format</p></td>
      </tr><tr><td>
          <p><code><a href="gintervals.convert_to_indexed.html">gintervals.convert_to_indexed()</a></code> </p>
        </td>
        <td><p>Convert 1D interval set to indexed format</p></td>
      </tr><tr><td>
          <p><code><a href="giterator.cartesian_grid.html">giterator.cartesian_grid()</a></code> </p>
        </td>
        <td><p>Creates a cartesian-grid iterator</p></td>
      </tr><tr><td>
          <p><code><a href="giterator.intervals.html">giterator.intervals()</a></code> </p>
        </td>
        <td><p>Returns iterator intervals</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data-manipulation-functions">Data manipulation functions <a href="#data-manipulation-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Data manipulation functions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gdist.html">gdist()</a></code> </p>
        </td>
        <td><p>Calculates distribution of track expressions</p></td>
      </tr><tr><td>
          <p><code><a href="gbins.quantiles.html">gbins.quantiles()</a></code> </p>
        </td>
        <td><p>Calculates quantiles of a track expression for bins</p></td>
      </tr><tr><td>
          <p><code><a href="gbins.summary.html">gbins.summary()</a></code> </p>
        </td>
        <td><p>Calculates summary statistics of a track expression for bins</p></td>
      </tr><tr><td>
          <p><code><a href="gcis_decay.html">gcis_decay()</a></code> </p>
        </td>
        <td><p>Calculates distribution of contact distances</p></td>
      </tr><tr><td>
          <p><code><a href="gcompute_strands_autocorr.html">gcompute_strands_autocorr()</a></code> </p>
        </td>
        <td><p>Computes auto-correlation between the strands for a file of mapped sequences</p></td>
      </tr><tr><td>
          <p><code><a href="gextract.html">gextract()</a></code> </p>
        </td>
        <td><p>Returns evaluated track expression</p></td>
      </tr><tr><td>
          <p><code><a href="glookup.html">glookup()</a></code> </p>
        </td>
        <td><p>Returns values from a lookup table based on track expression</p></td>
      </tr><tr><td>
          <p><code><a href="gpartition.html">gpartition()</a></code> </p>
        </td>
        <td><p>Partitions the values of track expression</p></td>
      </tr><tr><td>
          <p><code><a href="gquantiles.html">gquantiles()</a></code> </p>
        </td>
        <td><p>Calculates quantiles of a track expression</p></td>
      </tr><tr><td>
          <p><code><a href="gsample.html">gsample()</a></code> </p>
        </td>
        <td><p>Returns samples from the values of track expression</p></td>
      </tr><tr><td>
          <p><code><a href="gscreen.html">gscreen()</a></code> </p>
        </td>
        <td><p>Finds intervals that match track expression</p></td>
      </tr><tr><td>
          <p><code><a href="gsegment.html">gsegment()</a></code> </p>
        </td>
        <td><p>Divides track expression into segments</p></td>
      </tr><tr><td>
          <p><code><a href="gsummary.html">gsummary()</a></code> </p>
        </td>
        <td><p>Calculates summary statistics of track expression</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="sequence-functions">Sequence functions <a href="#sequence-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Sequence functions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gseq.extract.html">gseq.extract()</a></code> </p>
        </td>
        <td><p>Returns DNA sequences</p></td>
      </tr><tr><td>
          <p><code><a href="gseq.pwm.html">gseq.pwm()</a></code> </p>
        </td>
        <td><p>Score DNA sequences with a PWM over a region of interest</p></td>
      </tr><tr><td>
          <p><code><a href="gseq.kmer.html">gseq.kmer()</a></code> </p>
        </td>
        <td><p>Score DNA sequences with a k-mer over a region of interest</p></td>
      </tr><tr><td>
          <p><code><a href="gseq.revcomp.html">gseq.revcomp()</a></code> </p>
        </td>
        <td><p>Get reverse complement of DNA sequence</p></td>
      </tr><tr><td>
          <p><code><a href="grevcomp.html">grevcomp()</a></code> </p>
        </td>
        <td><p>Get reverse complement of DNA sequence</p></td>
      </tr><tr><td>
          <p><code><a href="gseq.rev.html">gseq.rev()</a></code> </p>
        </td>
        <td><p>Reverse DNA sequence</p></td>
      </tr><tr><td>
          <p><code><a href="gseq.comp.html">gseq.comp()</a></code> </p>
        </td>
        <td><p>Complement DNA sequence</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="database-functions">Database functions <a href="#database-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Database functions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gdb.create_genome.html">gdb.create_genome()</a></code> </p>
        </td>
        <td><p>Create and Load a Genome Database</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.create.html">gdb.create()</a></code> </p>
        </td>
        <td><p>Creates a new Genomic Database</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.get_readonly_attrs.html">gdb.get_readonly_attrs()</a></code> </p>
        </td>
        <td><p>Returns a list of read-only track attributes</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.init.html">gdb.init()</a></code> <code><a href="gdb.init.html">gdb.init_examples()</a></code> <code><a href="gdb.init.html">gsetroot()</a></code> </p>
        </td>
        <td><p>Initializes connection with Genomic Database</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.reload.html">gdb.reload()</a></code> </p>
        </td>
        <td><p>Reloads database from the disk</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.set_readonly_attrs.html">gdb.set_readonly_attrs()</a></code> </p>
        </td>
        <td><p>Sets read-only track attributes</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.convert_to_indexed.html">gdb.convert_to_indexed()</a></code> </p>
        </td>
        <td><p>Change Database to Indexed Genome Format</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.info.html">gdb.info()</a></code> </p>
        </td>
        <td><p>Get Database Information</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.mark_cache_dirty.html">gdb.mark_cache_dirty()</a></code> </p>
        </td>
        <td><p>Mark cached track list as dirty</p></td>
      </tr><tr><td>
          <p><code><a href="gdir.cd.html">gdir.cd()</a></code> </p>
        </td>
        <td><p>Changes current working directory in Genomic Database</p></td>
      </tr><tr><td>
          <p><code><a href="gdir.create.html">gdir.create()</a></code> </p>
        </td>
        <td><p>Creates a new directory in Genomic Database</p></td>
      </tr><tr><td>
          <p><code><a href="gdir.cwd.html">gdir.cwd()</a></code> </p>
        </td>
        <td><p>Returns the current working directory in Genomic Database</p></td>
      </tr><tr><td>
          <p><code><a href="gdir.rm.html">gdir.rm()</a></code> </p>
        </td>
        <td><p>Deletes a directory from Genomic Database</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="other-functions">Other functions <a href="#other-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Other</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="misha-package.html">misha-package</a></code> <code><a href="misha-package.html">misha</a></code> </p>
        </td>
        <td><p>Toolkit for analysis of genomic data</p></td>
      </tr><tr><td>
          <p><code><a href="gcluster.run.html">gcluster.run()</a></code> </p>
        </td>
        <td><p>Runs R commands on a cluster</p></td>
      </tr><tr><td>
          <p><code><a href="gwget.html">gwget()</a></code> </p>
        </td>
        <td><p>Downloads files from FTP server</p></td>
      </tr><tr><td>
          <p><code><a href="gwilcox.html">gwilcox()</a></code> </p>
        </td>
        <td><p>Calculates Wilcoxon test on sliding windows over track expression</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

