<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Toolkit for analysis of genomic data — misha-package • misha</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Toolkit for analysis of genomic data — misha-package"><meta property="og:description" content="'misha' package is intended to help users to efficiently analyze genomic
data achieved from various experiments."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Database-Formats.html">Database Formats and Multi-Contig Support</a>
    </li>
    <li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Toolkit for analysis of genomic data</h1>
    <small class="dont-index">Source: <a href="https://github.com/tanaylab/misha/blob/master/R/misha-package.R" class="external-link"><code>R/misha-package.R</code></a></small>
    <div class="hidden name"><code>misha-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>'misha' package is intended to help users to efficiently analyze genomic
data achieved from various experiments.</p>
    </div>


    <div id="details">
    <h2>Details</h2>
    <p>For a complete list of help resources, use <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library(help = "misha")</a></code>.</p>
<p>The following options are available for the package. Use 'options' function
to alter the value of the options.</p>
<table class="table table"><tr><td>NAME</td><td>DEFAULT</td><td>DESCRIPTION</td></tr><tr><td>gmax.data.size</td><td>AUTO</td><td>Auto-configured based on system RAM and processes.</td></tr><tr><td></td><td></td><td>Formula: min((RAM * 0.7) / gmax.processes, 10GB).</td></tr><tr><td></td><td></td><td>Controls
max in-memory result buffer size for streaming/sampling.</td></tr><tr><td></td><td></td><td>Operations like gextract, gscreen use this as upper bound.</td></tr><tr><td>gbig.intervals.size</td><td>1000000</td><td>Threshold for converting interval sets to disk-based</td></tr><tr><td></td><td></td><td>"big" format. When interval count exceeds this, intervals</td></tr><tr><td></td><td></td><td>are stored per-chromosome instead of all in memory.</td></tr><tr><td></td><td></td><td>Note: Independent of gmax.data.size (different purposes).</td></tr><tr><td>gmax.mem.usage</td><td>10000000</td><td>Maximal memory consumption of all child</td></tr><tr><td></td><td></td><td>processes in KB before the limiting algorithm is invoked.</td></tr><tr><td>gmax.processes</td><td>AUTO</td><td>Auto-configured to 70% of CPU cores.</td></tr><tr><td></td><td></td><td>Maximal number of processes for multitasking.</td></tr><tr><td>gmax.processes2core</td><td>2</td><td>Maximal number of processes per CPU core for multitasking</td></tr><tr><td>gmin.scope4process</td><td>10000</td><td>Minimal scope range (for 2D: surface)
assigned to a process</td></tr><tr><td></td><td></td><td>in multitasking mode.</td></tr><tr><td>gbuf.size</td><td>1000</td><td>Size of track expression values buffer.</td></tr><tr><td>gmultitask.max.records.factor</td><td>64</td><td>Multiplier to inflate multitask</td></tr><tr><td></td><td></td><td>max_records estimates to avoid under-allocation.</td></tr><tr><td>gtrack.chunk.size</td><td>100000</td><td>Chunk size in bytes of a 2D track. If '0' chunk size</td></tr><tr><td></td><td></td><td>is unlimited.</td></tr><tr><td>gtrack.num.chunks</td><td>0</td><td>Maximal number of 2D track
chunks simultaneously stored</td></tr><tr><td></td><td></td><td>in memory.</td></tr><tr><td>gmultitasking</td><td>TRUE</td><td>Enable/disable automatic parallelization. Some functions</td></tr><tr><td></td><td></td><td>may choose single-threaded mode for very small workloads.</td></tr></table><p>More information about the options can be found in 'User manual' of the
package.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://tanaylab.github.io/misha/" class="external-link">https://tanaylab.github.io/misha/</a></p></li>
<li><p><a href="https://github.com/tanaylab/misha" class="external-link">https://github.com/tanaylab/misha</a></p></li>
<li><p>Report bugs at <a href="https://github.com/tanaylab/misha/issues" class="external-link">https://github.com/tanaylab/misha/issues</a></p></li>
</ul></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p><strong>Maintainer</strong>: Aviezer Lifshitz <a href="mailto:aviezer.lifshitz@weizmann.ac.il">aviezer.lifshitz@weizmann.ac.il</a></p>
<p>Authors:</p><ul><li><p>Misha Hoichman <a href="mailto:misha@hoichman.com">misha@hoichman.com</a></p></li>
<li><p>Eitan Yaffe <a href="mailto:eitan.yaffe@weizmann.ac.il">eitan.yaffe@weizmann.ac.il</a></p></li>
<li><p>Amos Tanay <a href="mailto:amos.tanay@weizmann.ac.il">amos.tanay@weizmann.ac.il</a></p></li>
</ul><p>Other contributors:</p><ul><li><p>Weizmann Institute of Science [copyright holder]</p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

