<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ misha</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">misha</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/manual.Rnw">
    <span class="fa fa-stethoscope"></span>
     
    User Manual
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-th-list"></span>
     
    Reference
  </a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://bitbucket.org/aviezerl/misha-package">
    <span class="fa fa-bitbucket-square"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9">
    <div class="page-header">
      <h1>
        Reference
        <small>version&nbsp;3.7.1</small>
      </h1>
    </div>

    <div class="contents">
      <table class="ref-index">

      <colgroup>
        <col class="alias" />
        <col class="title" />
      </colgroup>

      <tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-track-functions" class="hasAnchor"><a href="#section-track-functions" class="anchor"></a>Track functions</h2>
            <p class="section-desc"><p>Functions to manipulate and create tracks</p></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.2d.create.html">gtrack.2d.create</a></code> </p>
          </td>
          <td><p>Creates a 'Rectangles' track from intervals and values</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.2d.import.html">gtrack.2d.import</a></code> </p>
          </td>
          <td><p>Creates a 2D track from tab-delimited file</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.2d.import_contacts.html">gtrack.2d.import_contacts</a></code> </p>
          </td>
          <td><p>Creates a track from a file of inter-genomic contacts</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.array.extract.html">gtrack.array.extract</a></code> </p>
          </td>
          <td><p>Returns values from 'Array' track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.array.get_colnames.html">gtrack.array.get_colnames</a></code> </p>
          </td>
          <td><p>Returns column names of array track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.array.import.html">gtrack.array.import</a></code> </p>
          </td>
          <td><p>Creates an array track from array tracks or files</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.array.set_colnames.html">gtrack.array.get_colnames</a></code> </p>
          </td>
          <td><p>Sets column names of array track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.attr.export.html">gtrack.attr.export</a></code> </p>
          </td>
          <td><p>Returns track attributes values</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.attr.get.html">gtrack.attr.get</a></code> </p>
          </td>
          <td><p>Returns value of a track attribute</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.attr.import.html">gtrack.attr.import</a></code> </p>
          </td>
          <td><p>Imports track attributes values</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.attr.set.html">gtrack.attr.set</a></code> </p>
          </td>
          <td><p>Assigns value to a track attribute</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.convert.html">gtrack.convert</a></code> </p>
          </td>
          <td><p>Converts a track to the most current format</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.create.html">gtrack.create</a></code> </p>
          </td>
          <td><p>Creates a track from a track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.create_pwm_energy.html">gtrack.create_pwm_energy</a></code> </p>
          </td>
          <td><p>Creates a new track from PSSM energy function</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.create_sparse.html">gtrack.create_sparse</a></code> </p>
          </td>
          <td><p>Creates a 'Sparse' track from intervals and values</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.exists.html">gtrack.exists</a></code> </p>
          </td>
          <td><p>Tests for a track existence</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.import.html">gtrack.import</a></code> </p>
          </td>
          <td><p>Creates a track from WIG / BigWig / BedGraph / tab-delimited file</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.import_mappedseq.html">gtrack.import_mappedseq</a></code> </p>
          </td>
          <td><p>Creates a track from a file of mapped sequences</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.import_set.html">gtrack.import_set</a></code> </p>
          </td>
          <td><p>Creates one or more tracks from multiple WIG / BigWig / BedGraph / tab-delimited files on disk or FTP</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.info.html">gtrack.info</a></code> </p>
          </td>
          <td><p>Returns information about a track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.liftover.html">gtrack.liftover</a></code> </p>
          </td>
          <td><p>Imports a track from another assembly</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.lookup.html">gtrack.lookup</a></code> </p>
          </td>
          <td><p>Creates a new track from a lookup table based on track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.ls.html">gtrack.ls</a></code> </p>
          </td>
          <td><p>Returns a list of track names</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.modify.html">gtrack.modify</a></code> </p>
          </td>
          <td><p>Modifies track contents</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.rm.html">gtrack.rm</a></code> </p>
          </td>
          <td><p>Deletes a track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.smooth.html">gtrack.smooth</a></code> </p>
          </td>
          <td><p>Creates a new track from smoothed values of track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.var.get.html">gtrack.var.get</a></code> </p>
          </td>
          <td><p>Returns value of a track variable</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.var.ls.html">gtrack.var.ls</a></code> </p>
          </td>
          <td><p>Returns a list of track variables for a track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.var.rm.html">gtrack.var.rm</a></code> </p>
          </td>
          <td><p>Deletes a track variable</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtrack.var.set.html">gtrack.var.set</a></code> </p>
          </td>
          <td><p>Assigns value to a track variable</p></td>
        </tr>
      </tbody><tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-virtual-track-functions" class="hasAnchor"><a href="#section-virtual-track-functions" class="anchor"></a>Virtual track functions</h2>
            <p class="section-desc"><p>Functions to manipulate and create virtual tracks</p></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="gvtrack.array.slice.html">gvtrack.array.slice</a></code> </p>
          </td>
          <td><p>Defines rules for a single value calculation of a virtual 'Array' track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gvtrack.create.html">gvtrack.create</a></code> </p>
          </td>
          <td><p>Creates a new virtual track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gvtrack.info.html">gvtrack.info</a></code> </p>
          </td>
          <td><p>Returns the definition of a virtual track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gvtrack.iterator.html">gvtrack.iterator</a></code> </p>
          </td>
          <td><p>Defines modification rules for a one-dimensional iterator in a virtual track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gvtrack.iterator.2d.html">gvtrack.iterator.2d</a></code> </p>
          </td>
          <td><p>Defines modification rules for a two-dimensional iterator in a virtual track</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gvtrack.ls.html">gvtrack.ls</a></code> </p>
          </td>
          <td><p>Returns a list of virtual track names</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gvtrack.rm.html">gvtrack.rm</a></code> </p>
          </td>
          <td><p>Deletes a virtual track</p></td>
        </tr>
      </tbody><tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-intervals-functions" class="hasAnchor"><a href="#section-intervals-functions" class="anchor"></a>Intervals functions</h2>
            <p class="section-desc"><p>Functions to manipulate and create intervals</p></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.html">gintervals</a></code> </p>
          </td>
          <td><p>Creates a set of 1D intervals</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.2d.html">gintervals.2d</a></code> </p>
          </td>
          <td><p>Creates a set of 2D intervals</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.2d.all.html">gintervals.2d.all</a></code> </p>
          </td>
          <td><p>Returns 2D intervals that cover the whole genome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.2d.band_intersect.html">gintervals.2d.band_intersect</a></code> </p>
          </td>
          <td><p>Intersects two-dimenstional intervals with a band</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.all.html">gintervals.all</a></code> </p>
          </td>
          <td><p>Returns 1D intervals that cover the whole genome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.canonic.html">gintervals.canonic</a></code> </p>
          </td>
          <td><p>Converts intervals to canonic form</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.chrom_sizes.html">gintervals.chrom_sizes</a></code> </p>
          </td>
          <td><p>Returns number of intervals per chromosome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.diff.html">gintervals.diff</a></code> </p>
          </td>
          <td><p>Calculates difference of two intervals sets</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.exists.html">gintervals.exists</a></code> </p>
          </td>
          <td><p>Tests for a named intervals set existence</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.force_range.html">gintervals.force_range</a></code> </p>
          </td>
          <td><p>Limits intervals to chromosomal range</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.import_genes.html">gintervals.import_genes</a></code> </p>
          </td>
          <td><p>Imports genes and annotations from files</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.intersect.html">gintervals.intersect</a></code> </p>
          </td>
          <td><p>Calculates an intersection of two sets of intervals</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.is.bigset.html">gintervals.is.bigset</a></code> </p>
          </td>
          <td><p>Tests for big intervals set</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.liftover.html">gintervals.liftover</a></code> </p>
          </td>
          <td><p>Converts intervals from another assembly</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.load.html">gintervals.load</a></code> </p>
          </td>
          <td><p>Loads a named intervals set</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.load_chain.html">gintervals.load_chain</a></code> </p>
          </td>
          <td><p>Loads assembly conversion table from a chain file</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.ls.html">gintervals.ls</a></code> </p>
          </td>
          <td><p>Returns a list of named intervals sets</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.mapply.html">gintervals.mapply</a></code> </p>
          </td>
          <td><p>Applies a function to values of track expressions</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.neighbors.html">gintervals.neighbors</a></code> </p>
          </td>
          <td><p>Finds neighbors between two sets of intervals</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.quantiles.html">gintervals.quantiles</a></code> </p>
          </td>
          <td><p>Calculates quantiles of a track expression for intervals</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.rbind.html">gintervals.rbind</a></code> </p>
          </td>
          <td><p>Combines several sets of intervals</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.rm.html">gintervals.rm</a></code> </p>
          </td>
          <td><p>Deletes a named intervals set</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.save.html">gintervals.save</a></code> </p>
          </td>
          <td><p>Creates a named intervals set</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.summary.html">gintervals.summary</a></code> </p>
          </td>
          <td><p>Calculates summary statistics of track expression for intervals</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.union.html">gintervals.union</a></code> </p>
          </td>
          <td><p>Calculates a union of two sets of intervals</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gintervals.update.html">gintervals.update</a></code> </p>
          </td>
          <td><p>Updates a named intervals set</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="giterator.cartesian_grid.html">giterator.cartesian_grid</a></code> </p>
          </td>
          <td><p>Creates a cartesian-grid iterator</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="giterator.intervals.html">giterator.intervals</a></code> </p>
          </td>
          <td><p>Returns iterator intervals</p></td>
        </tr>
      </tbody><tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-data-manipulation-functions" class="hasAnchor"><a href="#section-data-manipulation-functions" class="anchor"></a>Data manipulation functions</h2>
            <p class="section-desc"><p>Data manipulation functions</p></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="gdist.html">gdist</a></code> </p>
          </td>
          <td><p>Calculates distribution of track expressions</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gbins.quantiles.html">gbins.quantiles</a></code> </p>
          </td>
          <td><p>Calculates quantiles of a track expression for bins</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gbins.summary.html">gbins.summary</a></code> </p>
          </td>
          <td><p>Calculates summary statistics of a track expression for bins</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gcis_decay.html">gcis_decay</a></code> </p>
          </td>
          <td><p>Calculates distribution of contact distances</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gcompute_strands_autocorr.html">gcompute_strands_autocorr</a></code> </p>
          </td>
          <td><p>Computes auto-correlation between the strands for a file of mapped sequences</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gextract.html">gextract</a></code> </p>
          </td>
          <td><p>Returns evaluated track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="glookup.html">glookup</a></code> </p>
          </td>
          <td><p>Returns values from a lookup table based on track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gpartition.html">gpartition</a></code> </p>
          </td>
          <td><p>Partitions the values of track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gquantiles.html">gquantiles</a></code> </p>
          </td>
          <td><p>Calculates quantiles of a track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gsample.html">gsample</a></code> </p>
          </td>
          <td><p>Returns samples from the values of track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gscreen.html">gscreen</a></code> </p>
          </td>
          <td><p>Finds intervals that match track expression</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gsegment.html">gsegment</a></code> </p>
          </td>
          <td><p>Divides track expression into segments</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gseq.extract.html">gseq.extract</a></code> </p>
          </td>
          <td><p>Returns DNA sequences</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gsummary.html">gsummary</a></code> </p>
          </td>
          <td><p>Calculates summary statistics of track expression</p></td>
        </tr>
      </tbody><tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-database-functions" class="hasAnchor"><a href="#section-database-functions" class="anchor"></a>Database functions</h2>
            <p class="section-desc"><p>Database functions</p></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="gdb.create.html">gdb.create</a></code> </p>
          </td>
          <td><p>Creates a new Genomic Database</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gdb.get_readonly_attrs.html">gdb.get_readonly_attrs</a></code> </p>
          </td>
          <td><p>Returns a list of read-only track attributes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gdb.init.html">gdb.init</a></code> <code><a href="gdb.init.html">gdb.init_examples</a></code> </p>
          </td>
          <td><p>Initializes connection with Genomic Database</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gdb.reload.html">gdb.reload</a></code> </p>
          </td>
          <td><p>Reloads database from the disk</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gdb.set_readonly_attrs.html">gdb.set_readonly_attrs</a></code> </p>
          </td>
          <td><p>Sets read-only track attributes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gdir.cd.html">gdir.cd</a></code> </p>
          </td>
          <td><p>Changes current working directory in Genomic Database</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gdir.create.html">gdir.create</a></code> </p>
          </td>
          <td><p>Creates a new directory in Genomic Database</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gdir.cwd.html">gdir.cwd</a></code> </p>
          </td>
          <td><p>Returns the current working directory in Genomic Database</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gdir.rm.html">gdir.rm</a></code> </p>
          </td>
          <td><p>Deletes a directory from Genomic Database</p></td>
        </tr>
      </tbody><tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-other-functions" class="hasAnchor"><a href="#section-other-functions" class="anchor"></a>Other functions</h2>
            <p class="section-desc"><p>Other</p></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="misha-package.html">misha-package</a></code> </p>
          </td>
          <td><p>Toolkit for analysis of genomic data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gcluster.run.html">gcluster.run</a></code> </p>
          </td>
          <td><p>Runs R commands on a cluster</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gset_input_mode.html">gset_input_mode</a></code> </p>
          </td>
          <td><p>Sets input mode and auto-completion for track expressions, track names, virtual tracks and interval sets</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gtraceback.html">gtraceback</a></code> </p>
          </td>
          <td><p>Prints call stack of the last uncaught error</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gwget.html">gwget</a></code> </p>
          </td>
          <td><p>Downloads files from FTP server</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gwilcox.html">gwilcox</a></code> </p>
          </td>
          <td><p>Calculates Wilcoxon test on sliding windows over track expression</p></td>
        </tr>
      </tbody>
      </table>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-track-functions">Track functions</a></li>
      <li><a href="#section-virtual-track-functions">Virtual track functions</a></li>
      <li><a href="#section-intervals-functions">Intervals functions</a></li>
      <li><a href="#section-data-manipulation-functions">Data manipulation functions</a></li>
      <li><a href="#section-database-functions">Database functions</a></li>
      <li><a href="#section-other-functions">Other functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Misha Hoichman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
