<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Database Formats and Multi-Contig Support • misha</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Database Formats and Multi-Contig Support">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">misha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Database-Formats.html">Database Formats and Multi-Contig Support</a>
    </li>
    <li>
      <a href="../articles/Genomes.html">Genomes</a>
    </li>
    <li>
      <a href="../articles/Manual.html">Manual</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tanaylab/misha/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Database Formats and Multi-Contig Support</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/tanaylab/misha/blob/master/vignettes/Database-Formats.Rmd" class="external-link"><code>vignettes/Database-Formats.Rmd</code></a></small>
      <div class="hidden name"><code>Database-Formats.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tanaylab.github.io/misha/" class="external-link">misha</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Starting with misha 5.3.0, databases can be stored in two
formats:</p>
<ul>
<li>
<strong>Indexed format</strong> (default, recommended): Single
unified files for sequences and tracks</li>
<li>
<strong>Per-chromosome format</strong> (legacy): Separate files for
each chromosome</li>
</ul>
<p>The indexed format provides better performance and scalability,
especially for genomes with many contigs (&gt;50 chromosomes).</p>
<div class="section level3">
<h3 id="key-features">Key Features<a class="anchor" aria-label="anchor" href="#key-features"></a>
</h3>
<ul>
<li>
<strong>Automatic format detection</strong> - misha automatically
detects which format your database uses</li>
<li>
<strong>Fully backward compatible</strong> - existing databases
continue to work without modification</li>
<li>
<strong>Transparent to users</strong> - same API for both
formats</li>
<li>
<strong>Migration tools</strong> - convert databases when
convenient</li>
<li>
<strong>Performance benefits</strong> - 4-14% faster for large-scale
analyses</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="database-formats">Database Formats<a class="anchor" aria-label="anchor" href="#database-formats"></a>
</h2>
<div class="section level3">
<h3 id="indexed-format-recommended">Indexed Format (Recommended)<a class="anchor" aria-label="anchor" href="#indexed-format-recommended"></a>
</h3>
<p>The indexed format uses unified files:</p>
<p><strong>Sequence data:</strong> - <code>seq/genome.seq</code> - All
chromosome sequences concatenated - <code>seq/genome.idx</code> - Index
mapping chromosome names to positions</p>
<p><strong>Track data:</strong> -
<code>tracks/mytrack.track/track.dat</code> - All chromosome data
concatenated - <code>tracks/mytrack.track/track.idx</code> - Index with
offset/length per chromosome</p>
<p><strong>Advantages:</strong> - Fewer file descriptors (important for
genomes with 100+ contigs) - Better performance for large workloads (14%
faster) - Smaller disk footprint - Faster track creation and
conversion</p>
</div>
<div class="section level3">
<h3 id="per-chromosome-format-legacy">Per-Chromosome Format (Legacy)<a class="anchor" aria-label="anchor" href="#per-chromosome-format-legacy"></a>
</h3>
<p>The per-chromosome format uses separate files:</p>
<p><strong>Sequence data:</strong> - <code>seq/chr1.seq</code>,
<code>seq/chr2.seq</code>, … - One file per chromosome</p>
<p><strong>Track data:</strong> -
<code>tracks/mytrack.track/chr1.track</code>, <code>chr2.track</code>, …
- One file per chromosome</p>
<p><strong>When to use:</strong> - Compatibility with older misha
versions (&lt;5.3.0) - Small genomes (&lt;25 chromosomes) where
performance difference is negligible</p>
</div>
</div>
<div class="section level2">
<h2 id="creating-databases">Creating Databases<a class="anchor" aria-label="anchor" href="#creating-databases"></a>
</h2>
<div class="section level3">
<h3 id="new-databases-indexed-format">New Databases (Indexed Format)<a class="anchor" aria-label="anchor" href="#new-databases-indexed-format"></a>
</h3>
<p>By default, new databases use the indexed format:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create database from FASTA file</span></span>
<span><span class="fu"><a href="../reference/gdb.create.html">gdb.create</a></span><span class="op">(</span><span class="st">"mydb"</span>, <span class="st">"/path/to/genome.fa"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or download pre-built genome</span></span>
<span><span class="fu"><a href="../reference/gdb.create_genome.html">gdb.create_genome</a></span><span class="op">(</span><span class="st">"hg38"</span>, path <span class="op">=</span> <span class="st">"/path/to/install"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="force-legacy-format">Force Legacy Format<a class="anchor" aria-label="anchor" href="#force-legacy-format"></a>
</h3>
<p>To create a database in legacy format (for compatibility with older
misha):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set option before creating database</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>gmulticontig.indexed_format <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gdb.create.html">gdb.create</a></span><span class="op">(</span><span class="st">"mydb"</span>, <span class="st">"/path/to/genome.fa"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="checking-database-format">Checking Database Format<a class="anchor" aria-label="anchor" href="#checking-database-format"></a>
</h2>
<p>Use <code><a href="../reference/gdb.info.html">gdb.info()</a></code> to check your database format:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/mydb"</span><span class="op">)</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.info.html">gdb.info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">info</span><span class="op">$</span><span class="va">format</span><span class="op">)</span> <span class="co"># "indexed" or "per-chromosome"</span></span></code></pre></div>
<p>Example output:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.info.html">gdb.info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># $path</span></span>
<span><span class="co"># [1] "/path/to/mydb"</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $is_db</span></span>
<span><span class="co"># [1] TRUE</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $format</span></span>
<span><span class="co"># [1] "indexed"</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $num_chromosomes</span></span>
<span><span class="co"># [1] 24</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $genome_size</span></span>
<span><span class="co"># [1] 3095693983</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="converting-databases">Converting Databases<a class="anchor" aria-label="anchor" href="#converting-databases"></a>
</h2>
<div class="section level3">
<h3 id="convert-entire-database">Convert Entire Database<a class="anchor" aria-label="anchor" href="#convert-entire-database"></a>
</h3>
<p>Convert all tracks and sequences to indexed format:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/mydb"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gdb.convert_to_indexed.html">gdb.convert_to_indexed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This will: 1. Convert sequence files (<code>chr*.seq</code> →
<code>genome.seq + genome.idx</code>) 2. Convert all tracks to indexed
format 3. Validate conversions 4. Remove old files after successful
conversion</p>
</div>
<div class="section level3">
<h3 id="convert-individual-tracks">Convert Individual Tracks<a class="anchor" aria-label="anchor" href="#convert-individual-tracks"></a>
</h3>
<p>Convert specific tracks while keeping others in legacy format:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gtrack.convert_to_indexed.html">gtrack.convert_to_indexed</a></span><span class="op">(</span><span class="st">"mytrack"</span><span class="op">)</span></span></code></pre></div>
<p>Note that 2D tracks cannot be converted to indexed format yet.</p>
</div>
<div class="section level3">
<h3 id="convert-intervals">Convert Intervals<a class="anchor" aria-label="anchor" href="#convert-intervals"></a>
</h3>
<p>Convert interval sets to indexed format:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1D intervals</span></span>
<span><span class="fu"><a href="../reference/gintervals.convert_to_indexed.html">gintervals.convert_to_indexed</a></span><span class="op">(</span><span class="st">"myintervals"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2D intervals</span></span>
<span><span class="fu"><a href="../reference/gintervals.2d.convert_to_indexed.html">gintervals.2d.convert_to_indexed</a></span><span class="op">(</span><span class="st">"my2dintervals"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="migration-guide">Migration Guide<a class="anchor" aria-label="anchor" href="#migration-guide"></a>
</h2>
<div class="section level3">
<h3 id="when-to-migrate">When to Migrate<a class="anchor" aria-label="anchor" href="#when-to-migrate"></a>
</h3>
<p><strong>High priority (significant benefits):</strong> - Genomes with
many contigs (&gt;50 chromosomes) - Large-scale analyses (10M+ bp
regions frequently) - 2D track workflows - File descriptor limit
issues</p>
<p><strong>Medium priority (moderate benefits):</strong> - Repeated
extraction workflows - Regular analyses on medium-sized regions (1-10M
bp)</p>
<p><strong>Low priority (minimal benefits):</strong> - Small genomes
(&lt;25 chromosomes) - One-off analyses - Simple queries on small
regions</p>
</div>
<div class="section level3">
<h3 id="migration-workflow">Migration Workflow<a class="anchor" aria-label="anchor" href="#migration-workflow"></a>
</h3>
<p><strong>Step 1: Backup (optional but recommended)</strong></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create backup of important database</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"cp -r /path/to/mydb /path/to/mydb.backup"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 2: Check current format</strong></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/mydb"</span><span class="op">)</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.info.html">gdb.info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Current format:"</span>, <span class="va">info</span><span class="op">$</span><span class="va">format</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 3: Convert</strong></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdb.convert_to_indexed.html">gdb.convert_to_indexed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 4: Verify</strong></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check format changed</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.info.html">gdb.info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"New format:"</span>, <span class="va">info</span><span class="op">$</span><span class="va">format</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Test a few operations</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gextract.html">gextract</a></span><span class="op">(</span><span class="st">"mytrack"</span>, <span class="fu"><a href="../reference/gintervals.html">gintervals</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 5: Remove backup (after validation)</strong></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># After thorough testing</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"rm -rf /path/to/mydb.backup"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="copying-tracks-between-databases">Copying Tracks Between Databases<a class="anchor" aria-label="anchor" href="#copying-tracks-between-databases"></a>
</h2>
<p>You can freely copy tracks between databases with different
formats.</p>
<div class="section level3">
<h3 id="method-1-export-and-import">Method 1: Export and Import<a class="anchor" aria-label="anchor" href="#method-1-export-and-import"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Export from source database</span></span>
<span><span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/source_db"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gextract.html">gextract</a></span><span class="op">(</span><span class="st">"mytrack"</span>, <span class="fu"><a href="../reference/gintervals.all.html">gintervals.all</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    iterator <span class="op">=</span> <span class="st">"mytrack"</span>,</span>
<span>    file <span class="op">=</span> <span class="st">"/tmp/mytrack.txt"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import to target database (format auto-detected)</span></span>
<span><span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/target_db"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gtrack.import.html">gtrack.import</a></span><span class="op">(</span><span class="st">"mytrack"</span>, <span class="st">"Copied track"</span>, <span class="st">"/tmp/mytrack.txt"</span>, binsize <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co"># Automatically converted to target database format!</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="method-2-batch-copy">Method 2: Batch Copy<a class="anchor" aria-label="anchor" href="#method-2-batch-copy"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Copy multiple tracks</span></span>
<span><span class="va">tracks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"track1"</span>, <span class="st">"track2"</span>, <span class="st">"track3"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">track</span> <span class="kw">in</span> <span class="va">tracks</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Export</span></span>
<span>    <span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/source_db"</span><span class="op">)</span></span>
<span>    <span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"/tmp/%s.txt"</span>, <span class="va">track</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="../reference/gextract.html">gextract</a></span><span class="op">(</span><span class="va">track</span>, <span class="fu"><a href="../reference/gintervals.all.html">gintervals.all</a></span><span class="op">(</span><span class="op">)</span>, iterator <span class="op">=</span> <span class="va">track</span>, file <span class="op">=</span> <span class="va">file_path</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># Import</span></span>
<span>    <span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/target_db"</span><span class="op">)</span></span>
<span>    <span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gtrack.info.html">gtrack.info</a></span><span class="op">(</span><span class="va">track</span><span class="op">)</span> <span class="co"># Get description</span></span>
<span>    <span class="fu"><a href="../reference/gtrack.import.html">gtrack.import</a></span><span class="op">(</span><span class="va">track</span>, <span class="va">info</span><span class="op">$</span><span class="va">description</span>, <span class="va">file_path</span>, binsize <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">file_path</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="performance-comparison">Performance Comparison<a class="anchor" aria-label="anchor" href="#performance-comparison"></a>
</h2>
<p>Based on comprehensive benchmarks comparing indexed vs legacy
formats:</p>
<div class="section level3">
<h3 id="operations-faster-with-indexed-format">Operations Faster with Indexed Format<a class="anchor" aria-label="anchor" href="#operations-faster-with-indexed-format"></a>
</h3>
<ul>
<li>
<strong>Very large workloads (10M+ bp)</strong>: 14% faster</li>
<li>
<strong>2D track operations</strong>: 2% faster (was 36% slower in
legacy)</li>
<li>
<strong>Repeated extractions</strong>: 14% faster</li>
<li>
<strong>Real workflows</strong>: 8% faster on average</li>
</ul>
</div>
<div class="section level3">
<h3 id="operations-similar-performance">Operations Similar Performance<a class="anchor" aria-label="anchor" href="#operations-similar-performance"></a>
</h3>
<ul>
<li>Single chromosome extraction: Within 5%</li>
<li>Multi-chromosome (10-22 chr): Within 1%</li>
<li>PWM operations: Within 3%</li>
<li>Small workloads (&lt;1M bp): Within 10%</li>
</ul>
</div>
<div class="section level3">
<h3 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h3>
<ul>
<li>
<strong>64% of operations are faster</strong> with indexed
format</li>
<li>
<strong>93% within 5%</strong> of legacy format (statistically
equal)</li>
<li>
<strong>Average: 4% faster</strong> across all benchmarks</li>
<li>
<strong>No regressions</strong> for common use cases</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="backward-compatibility">Backward Compatibility<a class="anchor" aria-label="anchor" href="#backward-compatibility"></a>
</h2>
<div class="section level3">
<h3 id="fully-compatible">Fully Compatible<a class="anchor" aria-label="anchor" href="#fully-compatible"></a>
</h3>
<ul>
<li>✅ Existing databases work without modification</li>
<li>✅ Existing scripts work without changes</li>
<li>✅ Same API for both formats</li>
<li>✅ Automatic format detection</li>
<li>✅ Can mix formats in same analysis</li>
</ul>
</div>
<div class="section level3">
<h3 id="example-mixed-environment">Example: Mixed Environment<a class="anchor" aria-label="anchor" href="#example-mixed-environment"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Work with both formats in same session</span></span>
<span><span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/legacy_db"</span><span class="op">)</span></span>
<span><span class="va">data1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gextract.html">gextract</a></span><span class="op">(</span><span class="st">"track1"</span>, <span class="fu"><a href="../reference/gintervals.html">gintervals</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/gdb.init.html">gsetroot</a></span><span class="op">(</span><span class="st">"/path/to/indexed_db"</span><span class="op">)</span></span>
<span><span class="va">data2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gextract.html">gextract</a></span><span class="op">(</span><span class="st">"track2"</span>, <span class="fu"><a href="../reference/gintervals.html">gintervals</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="troubleshooting">Troubleshooting<a class="anchor" aria-label="anchor" href="#troubleshooting"></a>
</h2>
<div class="section level3">
<h3 id="file-descriptor-limit-reached">“File descriptor limit reached”<a class="anchor" aria-label="anchor" href="#file-descriptor-limit-reached"></a>
</h3>
<p>This occurs with many-contig genomes in legacy format:</p>
<p><strong>Solution:</strong> Convert to indexed format</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdb.convert_to_indexed.html">gdb.convert_to_indexed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="track-not-found-after-copying-files">“Track not found after copying files”<a class="anchor" aria-label="anchor" href="#track-not-found-after-copying-files"></a>
</h3>
<p>After manually copying track directories:</p>
<p><strong>Solution:</strong> Reload database</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdb.reload.html">gdb.reload</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="conversion-fails-with-disk-space-error">“Conversion fails with disk space error”<a class="anchor" aria-label="anchor" href="#conversion-fails-with-disk-space-error"></a>
</h3>
<p>Conversion needs 2x track size temporarily:</p>
<p><strong>Solution:</strong> Free disk space or convert tracks
individually</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert one track at a time</span></span>
<span><span class="fu"><a href="../reference/gtrack.convert_to_indexed.html">gtrack.convert_to_indexed</a></span><span class="op">(</span><span class="st">"track1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gtrack.convert_to_indexed.html">gtrack.convert_to_indexed</a></span><span class="op">(</span><span class="st">"track2"</span><span class="op">)</span></span>
<span><span class="co"># etc.</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="best-practices">Best Practices<a class="anchor" aria-label="anchor" href="#best-practices"></a>
</h2>
<div class="section level3">
<h3 id="for-new-projects">For New Projects<a class="anchor" aria-label="anchor" href="#for-new-projects"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Use indexed format (default) for new databases</li>
<li>Use <code><a href="../reference/gdb.create_genome.html">gdb.create_genome()</a></code> for standard genomes</li>
<li>Use <code><a href="../reference/gdb.create.html">gdb.create()</a></code> with multi-FASTA for custom
genomes</li>
</ol>
</div>
<div class="section level3">
<h3 id="for-existing-projects">For Existing Projects<a class="anchor" aria-label="anchor" href="#for-existing-projects"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Check format with <code><a href="../reference/gdb.info.html">gdb.info()</a></code>
</li>
<li>Migrate if beneficial (many contigs, large analyses)</li>
<li>No rush - legacy format remains fully supported</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="summary-1">Summary<a class="anchor" aria-label="anchor" href="#summary-1"></a>
</h2>
<ul>
<li>
<strong>Indexed format is the default and recommended</strong> for
new databases</li>
<li>
<strong>Legacy format remains fully supported</strong> - no forced
migration</li>
<li>
<strong>Automatic format detection</strong> - users don’t need to
think about it</li>
<li>
<strong>Significant performance benefits</strong> for large-scale
analyses</li>
<li>
<strong>Easy migration</strong> with
<code><a href="../reference/gdb.convert_to_indexed.html">gdb.convert_to_indexed()</a></code>
</li>
<li>
<strong>Fully backward compatible</strong> - existing code works
unchanged</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Misha Hoichman, Aviezer Lifshitz, Eitan Yaffe, Amos Tanay.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
